lib/core/include/version.hpp

```diff
@@ -0,0 +1,90 @@
+#ifndef IRODS_VERSION_HPP
+#define IRODS_VERSION_HPP
+
+#include <cstdint>
+#include <functional>
+
+namespace irods
+{
+    /// A simple struct capable of representing a version number.
+    ///
+    /// \since 4.2.9
+    struct version
+    {
+        // clang-format off
+        std::uint16_t major    = 0;
+        std::uint16_t minor    = 0;
+        std::uint16_t patch    = 0;
+        std::uint16_t revision = 0;
```

Comment:
I always used the trailing underscore to signal to the reader that the member variable is private to the class/struct. They don't have a trailing underscore because these variables are part of the public interface. Willing to discuss more.